### PR TITLE
feat(spark): support spark-sink connector for loader

### PR DESCRIPTION
**Pull request by liuxiaocs7**
Another separate spark-module for hugegraph to read/write data

Ready to review for spark submodule 🤔 

TODO：add basic context for it & diff with #283

### CODE REVIEW

**Comment by javeme** [comment]
Is it still in progress?

**Comment by liuxiaocs7** [comment]
> Is it still in progress?

Sorry for late, I will continue to finish

File: hugegraph-spark-connector/src/main/scala/org/apache/hugegraph/spark/connector/utils/HGBuildUtils.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  def buildVertices(row: InternalRow, schema: StructType, builder: VertexBuilder): List[Vertex] = {
+    val fields = schema.names
+    val dataTypes = schema.fields.map(field => field.dataType)
+    val values = for {
+      idx <- schema.fields.indices
+    } yield {
+      val value = row.get(idx, dataTypes(idx))
+      if (value.getClass.getSimpleName.equalsIgnoreCase("UTF8String")) value.toString
+      else value
```
Comment (simon824) [inline hugegraph-spark-connector/src/main/scala/org/apache/hugegraph/spark/connector/utils/HGBuildUtils.scala:41]: ensure uniform code style and improve readability
```suggestion
      if (value.getClass.getSimpleName.equalsIgnoreCase("UTF8String")) {
        value.toString
      } else {
        value
      }
```

File: hugegraph-spark-connector/src/main/scala/org/apache/hugegraph/spark/connector/writer/HGBatchWriterFactory.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+class HGBatchWriterFactory(schema: StructType, hgOptions: HGOptions) extends DataWriterFactory {
+
+  private val LOG = LoggerFactory.getLogger(this.getClass)
+
+  override def createWriter(partitionId: Int, taskId: Long): DataWriter[InternalRow] = {
+    val dataType = hgOptions.dataType()
+    LOG.info(s"Create a ${dataType} writer, partitionId: ${partitionId}, taskId: ${taskId}")
+    if (dataType == "vertex") new HGVertexWriter(schema, hgOptions)
+    else new HGEdgeWriter(schema, hgOptions)
```
Comment (simon824) [inline hugegraph-spark-connector/src/main/scala/org/apache/hugegraph/spark/connector/writer/HGBatchWriterFactory.scala:35]: ditto

File: hugegraph-spark-connector/src/main/scala/org/apache/hugegraph/spark/connector/writer/HGEdgeWriter.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    if (edgesBuffer.size >= hgOptions.batchSize()) {
+      sinkOnce()
+    }
+  }
+
+  private def sinkOnce(): Unit = {
+    LOG.info(s"Writer once: ${edgesBuffer.toList}")
+    val successfulVertices = HGBuildUtils.saveEdges(context, edgesBuffer.toList)
+    val successIds = successfulVertices.map(v => v.id())
+    LOG.info(s"Successful ids: ${successIds}")
```
[the remark this answers is not in the document]
Comment (liuxiaocs7) [reply hugegraph-spark-connector/src/main/scala/org/apache/hugegraph/spark/connector/writer/HGEdgeWriter.scala]: fixed

File: hugegraph-spark-connector/src/main/java/org/apache/hugegraph/spark/connector/builder/EdgeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+import org.apache.hugegraph.structure.graph.Vertex;
+import org.apache.hugegraph.structure.schema.EdgeLabel;
+import org.apache.hugegraph.structure.schema.SchemaLabel;
+import org.apache.hugegraph.structure.schema.VertexLabel;
+import org.apache.hugegraph.util.E;
+
+import com.google.common.collect.ImmutableList;
+
+public class EdgeBuilder extends ElementBuilder<Edge> {
+
```
[the remark this answers is not in the document]
Comment (liuxiaocs7) [reply hugegraph-spark-connector/src/main/java/org/apache/hugegraph/spark/connector/builder/EdgeBuilder.java:39]: Yes, but the Mapping and Builder here removes some of the settings and the original loader contains too many dependencies, many of which are unnecessary.

The Java modules in this project can also be reused by different data integration tools.

File: hugegraph-spark-connector/src/test/java/org/apache/hugegraph/spark/connector/utils/HGEnvUtils.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    public static final String DEFAULT_HOST = "127.0.0.1";
+    public static final String DEFAULT_PORT = "8080";
+    public static final String DEFAULT_GRAPH = "hugegraph";
+
+    private static HugeClient hugeClient;
+
+    public static void createEnv() {
+
+        hugeClient = HugeClient.builder(
+                "http://" + DEFAULT_HOST + ":" + DEFAULT_PORT, DEFAULT_GRAPH).build();
```
Comment (javeme) [inline hugegraph-spark-connector/src/test/java/org/apache/hugegraph/spark/connector/utils/HGEnvUtils.java:34]: also define a URL const?

**Review by javeme — DISMISSED** [review]
LGTM

File: hugegraph-spark-connector/README.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+| `batch-size`      | `500`         | The number of data items in each batch when importing data                                                                                                                                                                                                                                                                                                                                                                 |
+
+### Common Configs
+
+| Params      | Default Value | Description                                                                     |
+|-------------|---------------|---------------------------------------------------------------------------------|
+| `delimiter` | `,`           | Separator of `source-name`, `target-name`, `selected-fields` or `ignore-fields` |
+
+## Example
```
Comment (javeme) [inline hugegraph-spark-connector/README.md:57]: prefer to put it at the beginning and rename to how to use, please outline what steps are required in total and which steps they are.

File: hugegraph-spark-connector/README.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+License for the specific language governing permissions and limitations
+under the License.
+-->
+
+# HugeGraph Spark Connector
+
```
Comment (javeme) [inline hugegraph-spark-connector/README.md:19]: can we add a section to describe what the project is and in what scenarios it is used.

File: hugegraph-spark-connector/src/test/java/org/apache/hugegraph/spark/connector/utils/HGEnvUtils.java (hunk 5 above)
Comment (liuxiaocs7) [reply hugegraph-spark-connector/src/test/java/org/apache/hugegraph/spark/connector/utils/HGEnvUtils.java]: done~

File: pom.xml [hunk 8]
```diff
@@ -93,6 +93,7 @@
         <module>hugegraph-tools</module>
         <module>hugegraph-hubble</module>
         <module>hugegraph-dist</module>
+        <module>hugegraph-spark-connector</module>
```
Comment (imbajin) [inline pom.xml:96]: the spark module is the LAST order in building?

File: .github/workflows/spark-connector-ci.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+      - name: Cache Maven packages
+        uses: actions/cache@v3
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
+
+      - name: Checkout
+        uses: actions/checkout@v3
```
Comment (imbajin) [inline .github/workflows/spark-connector-ci.yml:42]:
```suggestion
        uses: actions/checkout@v4
```

File: hugegraph-spark-connector/pom.xml [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        <groupId>org.apache.hugegraph</groupId>
+        <artifactId>hugegraph-toolchain</artifactId>
+        <version>${revision}</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>hugegraph-spark-connector</artifactId>
+    <packaging>jar</packaging>
+
+    <name>${project.artifactId}</name>
```
Comment (imbajin) [inline hugegraph-spark-connector/pom.xml:32]:
```suggestion
    <name>${project.artifactId}</name>
```

File: hugegraph-spark-connector/pom.xml [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+
+    <properties>
+        <maven.compiler.source>8</maven.compiler.source>
+        <maven.compiler.target>8</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <spark.verion>3.2.2</spark.verion>
+        <spark.scope>provided</spark.scope>
+        <hugegraph.client.version>1.0.0</hugegraph.client.version>
+        <hugegraph.common.version>1.0.0</hugegraph.common.version>
```
Comment (imbajin) [inline hugegraph-spark-connector/pom.xml:41]: client includes common, seems we should include one of them

File: hugegraph-spark-connector/README.md (hunk 7 above)
Comment (liuxiaocs7) [reply hugegraph-spark-connector/README.md:19]: Done~

**Comment by liuxiaocs7** [comment]
Hi, @javeme, @imbajin, @simon824, PTAL when free, thanks!

A simple example use this module: https://github.com/liuxiaocs7/HGSparkConnectorExample

**Review by simon824 — APPROVED** [review]
LGTM

**Comment by JackyYangPassion** [comment]
> Hi, @javeme, @imbajin, @simon824, PTAL when free, thanks!
> 
> A simple example use this module: [liuxiaocs7/HGSparkConnectorExample](https://github.com/liuxiaocs7/HGSparkConnectorExample)
> 
> The job:
> 
> image
> 
> Result:
> 
> image

Does this feature support pyspark reference operation?
like this
```
/spark/bin/pyspark --driver-class-path hugegraph-spark-connector-1.0.0.jar --jars hugegraph-spark-connector-1.0.0.jar

```

**Review by javeme — APPROVED** [review]
LGTM

**Review by imbajin — APPROVED** [review]
THX, remember add basic context/usage doc later~

**Pull request merged**